File: StriverA2Z/Java/LinkedLists/Questions/SinglyLinkedList/DeleteMiddle.java
public class DeleteMiddle {
    public static void main(String[] args) {
        Node head = LinkedListOperations.arrayToLinkedList(new int[] {1,2,3,4,5,6});
        head = deleteMiddle(head);
        LinkedListOperations.traverseLinkedList(head);
    }

    static Node deleteMiddle(Node head) {
        if (head == null || head.next == null) return null;
        Node fast = head;
        Node slow = head;
        fast = fast.next.next;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        slow.next = slow.next.next;
        return head;
    }
}

File: StriverA2Z/C++/6_LinkedList/Questions/SinglyLinkedList/delete_middle.cpp
#include<bits/stdc++.h>
#include "Node.h"

using namespace std;

Node *deleteMiddle(Node *head) {
    if (head == nullptr || head->next == nullptr) return nullptr;
    Node *slow = head;
    Node *fast = head;
    fast = fast->next->next;
    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
    }
    Node *middle = slow->next;
    slow->next = slow->next->next;
    delete middle;
    return head;
}

int main() {
    vector<int> arr = {1, 2, 3, 5, 6};
    Node *head = array_to_linkedlist(arr);
    head = deleteMiddle(head);
    traverseLinkedList(head);
    return 0;
}